File: src/api/homepage.ts
import { RequestFN } from "~/core/request-function";
import { encodePronoteDate } from "~/encoders/pronote-date";
import { type SessionHandle } from "~/models";
import { translateToWeekNumber } from "./helpers/week-number";
import { decodeHomepage } from "~/decoders/homepage";
import type { Homepage } from "~/models/homepage";

/**
 * Retrieve data from the homepage for the given session.
 */
export const homepage = async (session: SessionHandle, day = session.instance.nextBusinessDay): Promise<Homepage> => {
  const weekNumber = translateToWeekNumber(day, session.instance.firstMonday);
  const next = encodePronoteDate(day);

  const request = new RequestFN(session, "PageAccueil", {
    _Signature_: {
      onglet: 7 // = Presence
    },

    donnees: {
      avecConseilDeClasse: true,

      dateGrille: {
        _T: 7,
        V: next
      },

      numeroSemaine: weekNumber,

      coursNonAssures: {
        numeroSemaine: weekNumber
      },

      personnelsAbsents: {
        numeroSemaine: weekNumber
      },

      incidents: {
        numeroSemaine: weekNumber
      },

      exclusions: {
        numeroSemaine: weekNumber
      },

      donneesVS: {
        numeroSemaine: weekNumber
      },

      registreAppel: {
        date: {
          _T: 7,
          V: next
        }
      },

      previsionnelAbsServiceAnnexe: {
        date: {
          _T: 7,
          V: next
        }
      },

      donneesProfs: {
        numeroSemaine: weekNumber
      },

      EDT: {
        numeroSemaine: weekNumber
      },

      menuDeLaCantine: {
        date: {
          _T: 7,
          V: next
        }
      },

      TAFARendre: {
        date: {
          _T: 7,
          V: next
        }
      },

      TAFEtActivites: {
        date: {
          _T: 7,
          V: next
        }
      },

      partenaireCDI: {
        CDI: {}
      },

      tableauDeBord: {
        date: {
          _T: 7,
          V: next
        }
      },

      modificationsEDT: {
        date: {
          _T: 7,
          V: next
        }
      }
    }
  });

  const response = await request.send();
  return decodeHomepage(response.data.donnees);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/homepage.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.api.helpers.translateToWeekNumber
import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeHomepage
import ink.literate.pawnote.encoders.encodePronoteDate
import ink.literate.pawnote.models.Homepage
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun homepage(
    session: SessionHandle,
    day: LocalDateTime = session.instance.nextBusinessDay
): Homepage {
  val weekNumber = translateToWeekNumber(day, session.instance.firstMonday)
  val next = encodePronoteDate(day)

  val request =
      RequestFN(
          session.information,
          "PageAccueil",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Presence.code) }

                putJsonObject("donnees") {
                  put("avecConseilDeClasse", true)

                  putJsonObject("dateGrille") {
                    put("_T", 7)
                    put("V", next)
                  }

                  put("numeroSemaine", weekNumber)

                  putJsonObject("coursNonAssures") { put("numeroSemaine", weekNumber) }

                  putJsonObject("personnelsAbsents") { put("numeroSemaine", weekNumber) }

                  putJsonObject("incidents") { put("numeroSemaine", weekNumber) }

                  putJsonObject("exclusions") { put("numeroSemaine", weekNumber) }

                  putJsonObject("donneesVS") { put("numeroSemaine", weekNumber) }

                  putJsonObject("registreAppel") {
                    putJsonObject("date") {
                      put("_T", 7)
                      put("V", next)
                    }
                  }

                  putJsonObject("previsionnelAbsServiceAnnexe") {
                    putJsonObject("date") {
                      put("_T", 7)
                      put("V", next)
                    }
                  }

                  putJsonObject("donneesProfs") { put("numeroSemaine", weekNumber) }

                  putJsonObject("EDT") { put("numeroSemaine", weekNumber) }

                  putJsonObject("TAFARendre") {
                    putJsonObject("date") {
                      put("_T", 7)
                      put("V", next)
                    }
                  }

                  putJsonObject("partenaireCDI") { putJsonObject("CDI") {} }

                  putJsonObject("tableauDeBord") {
                    putJsonObject("date") {
                      put("_T", 7)
                      put("V", next)
                    }
                  }

                  putJsonObject("modificationsEDT") {
                    putJsonObject("date") {
                      put("_T", 7)
                      put("V", next)
                    }
                  }
                }
              }))

  val response = request.send()
  return decodeHomepage(
      Json.decodeFromString<JsonObject>(response.data).jsonObject["donnees"]!!.jsonObject)
}
